reana_commons/consumer.py

```diff
@@ -21,51 +21,50 @@
-        self._params = pika.connection.ConnectionParameters(
-            BROKER_URL, BROKER_PORT, '/', self.broker_credentials)
-        self._conn = None
-        self._channel = None
-        self._queue = queue
+    def __init__(self, connection=None, queues=None, default_serializer=None):
+        """Construct a REANAConsumer.
 
-    def connect(self):
-        """Connect to MQ channel."""
-        if not self._conn or self._conn.is_closed:
-            self._conn = pika.BlockingConnection(self._params)
-            self._channel = self._conn.channel()
-            self._channel.basic_qos(prefetch_count=1)
-            self._channel.queue_declare(queue=self._queue)
+        :param connection: kombu.Connection object.
+        :param queues: List of queues names the consumer will subscribe to.
+        """
+        self.queues = queues or self._build_default_queues()
+        self.connection = connection or Connection(BROKER)
+        self.default_serializer = default_serializer or MQ_DEFAULT_SERIALIZER
 
-    def on_message(self, channel, method_frame, header_frame, body):
-        """On new message event handler."""
-        # this method should be overriden by the inheriting class
-        pass
+    def _build_default_exchange(self):
+        """Build ``kombu.Exchange`` with default values."""
+        exchange = Exchange(MQ_DEFAULT_EXCHANGE, type='direct')
+        return exchange
 
-    def consume(self):
-        """Start consuming incoming messages."""
-        while True:
-            self.connect()
-            self._channel.basic_consume(self.on_message, self._queue)
-            try:
-                self._channel.start_consuming()
-            except KeyboardInterrupt:
-                self._channel.stop_consuming()
-                self._conn.close()
-            except pika.exceptions.ConnectionClosedByBroker:
-                # Uncomment this to make the example not attempt recovery
-                # from server-initiated connection closure, including
-                # when the node is stopped cleanly
-                # except pika.exceptions.ConnectionClosedByBroker:
-                #     pass
-                continue
+    def _build_default_queues(self):
+        """Build ``kombu.Queue`` with default values."""
+        default_queue = Queue(MQ_DEFAULT_QUEUE,
+                              durable=False,
+                              exchange=self._build_default_exchange(),
+                              routing_key=MQ_DEFAULT_ROUTING_KEY)
+        return [default_queue]
+
+    def get_consumers(self, Consumer, channel):
+        """Implement providing ``kombu.Consumer`` with queues/callbacks."""
+        # return [Consumer(queues=self.queues,
+        #                  callbacks=[self.on_message],
+        #                  accept=[MQ_DEFAULT_SERIALIZER]))]
+        raise NotImplementedError('Implement this method to map which'
+                                  'callbacks are connected with which queues')
```

Comment:
I don't understand this part very well. Would it be setting the consumers for specific queues?

Reply:
Yes, since the `ConsumerMixin` class is handling the reconnections it needs to know how to rebuild the `kombu.Consumer`'s. So here we build the consumer providing the queues where to consume from and the tasks to execute when consuming from there (without taking into account what they are since we have only one queue).

This is fine for now, but once we start to have different queues we will need different consumers (pairs `queue` - `callback`). I didn't get deep into this since I wanted to keep it simple for now, but bloating the same class with many callbacks and internally mapping them to the queues is indeed complicated, but we can iterate over it once we have more queues?

Reply:
Ok, I see. Yes, for now I would change the method signature so that it mentions a single consumer maybe. We could rephrase the error message to something simpler. It is understandable, just not clear when combined with the method name and parameters.